File: examples/end2end-example/rust/PriceBet_test.rs
#[path = "PriceBet.runar.rs"]
mod contract;

use contract::*;
use runar::prelude::*;

fn new_price_bet() -> PriceBet {
    PriceBet {
        alice_pub_key: ALICE.pub_key.to_vec(),
        bob_pub_key: BOB.pub_key.to_vec(),
        oracle_pub_key: oracle::pubkey_le(),
        strike_price: 50000,
    }
}

/// Honestly sign a price with the shared test Rabin key. Mirrors
/// `examples/end2end-example/go/PriceBet_test.go::signPrice` (which uses
/// `runar.RabinSignToBytes`) and the TS test's `rabinSign(RABIN_TEST_KEY)`.
/// `verify_rabin_sig` performs REAL Rabin verification (with the on-chain
/// padding bound 0 <= padding < 65536), so garbage (sig, padding) pairs no
/// longer verify.
fn sign_price(price: Bigint) -> (Vec<u8>, Vec<u8>) {
    let msg = num2bin(&price, 8);
    oracle::sign(&msg)
}

#[test]
fn test_settle_alice_wins() {
    let (rabin_sig, pad) = sign_price(60000);
    let alice_sig = ALICE.sign_test_message();
    let bob_sig = BOB.sign_test_message();
    new_price_bet().settle(60000, &rabin_sig, &pad, &alice_sig, &bob_sig);
}

#[test]
fn test_settle_bob_wins() {
    let (rabin_sig, pad) = sign_price(30000);
    let alice_sig = ALICE.sign_test_message();
    let bob_sig = BOB.sign_test_message();
    new_price_bet().settle(30000, &rabin_sig, &pad, &alice_sig, &bob_sig);
}

#[test]
fn test_settle_bob_wins_at_strike() {
    let (rabin_sig, pad) = sign_price(50000);
    let alice_sig = ALICE.sign_test_message();
    let bob_sig = BOB.sign_test_message();
    new_price_bet().settle(50000, &rabin_sig, &pad, &alice_sig, &bob_sig);
}

#[test]
#[should_panic(expected = "price > 0")]
fn test_settle_zero_price_rejected() {
    // Honestly signed so the Rabin layer passes and the failure is
    // attributable to the `price > 0` assert, not the signature check.
    let (rabin_sig, pad) = sign_price(0);
    let alice_sig = ALICE.sign_test_message();
    let bob_sig = BOB.sign_test_message();
    new_price_bet().settle(0, &rabin_sig, &pad, &alice_sig, &bob_sig);
}

#[test]
#[should_panic(expected = "verify_rabin_sig")]
fn test_settle_forged_oracle_sig_rejected() {
    // Price favors Alice but the oracle signature is garbage — the Rabin
    // layer must reject it.
    let alice_sig = ALICE.sign_test_message();
    let bob_sig = BOB.sign_test_message();
    let forged_sig: RabinSig = b"not_a_signature".to_vec();
    let pad: ByteString = vec![0u8];
    new_price_bet().settle(60000, &forged_sig, &pad, &alice_sig, &bob_sig);
}

#[test]
fn test_cancel() {
    let alice_sig = ALICE.sign_test_message();
    let bob_sig = BOB.sign_test_message();
    new_price_bet().cancel(&alice_sig, &bob_sig);
}

#[test]
fn test_compile() {
    runar::compile_check(
        include_str!("PriceBet.runar.rs"),
        "PriceBet.runar.rs",
    ).unwrap();
}

/// Test-only honest Rabin oracle. Same p, q ≡ 3 (mod 4) test key as the Go
/// (`runar.RabinTestP/Q`), Java (`OraclePriceFeedTest`), and integration-test
/// (`integration/rust/tests/helpers/crypto.rs`) tiers. Square roots mod p are
/// computable as a^((p+1)/4) mod p because p ≡ 3 (mod 4).
mod oracle {
    use num_bigint::BigUint;

    const P_DEC: &str = "1361129467683753853853498429727072846227";
    const Q_DEC: &str = "1361129467683753853853498429727082846007";

    fn p() -> BigUint {
        P_DEC.parse().unwrap()
    }

    fn q() -> BigUint {
        Q_DEC.parse().unwrap()
    }

    /// Rabin modulus n = p * q as unsigned LE bytes (the contract's RabinPubKey).
    pub fn pubkey_le() -> Vec<u8> {
        (p() * q()).to_bytes_le()
    }

    /// sqrt of `a` mod prime `p` where p ≡ 3 (mod 4); None if `a` is not a
    /// quadratic residue.
    fn sqrt_mod(a: &BigUint, p: &BigUint) -> Option<BigUint> {
        let a = a % p;
        if a == BigUint::from(0u8) {
            return Some(a);
        }
        let r = a.modpow(&((p + 1u8) >> 2), p);
        if (&r * &r) % p == a {
            Some(r)
        } else {
            None
        }
    }

    /// Honest Rabin signature: returns (sig, padding) as unsigned LE bytes
    /// with padding < 65536 such that (sig² + padding) mod n == SHA256(msg)
    /// mod n. Finds the padding by quadratic-residue search and the root via
    /// CRT (modular inverses by Fermat, since p and q are prime).
    pub fn sign(msg: &[u8]) -> (Vec<u8>, Vec<u8>) {
        let (p, q) = (p(), q());
        let n = &p * &q;
        let hash = runar::prelude::sha256(msg);
        let hash_mod_n = BigUint::from_bytes_le(&hash) % &n;
        let q_inv_p = q.modpow(&(&p - 2u8), &p);
        let p_inv_q = p.modpow(&(&q - 2u8), &q);
        for pad in 0u32..65536 {
            let pad_bn = BigUint::from(pad);
            let target = (&hash_mod_n + &n - &pad_bn) % &n;
            let (Some(rp), Some(rq)) = (sqrt_mod(&target, &p), sqrt_mod(&target, &q)) else {
                continue;
            };
            let sig = (&rp * &q * &q_inv_p + &rq * &p * &p_inv_q) % &n;
            if (&sig * &sig + &pad_bn) % &n == hash_mod_n {
                return (sig.to_bytes_le(), pad_bn.to_bytes_le());
            }
        }
        panic!("no valid padding < 65536 found for test message");
    }
}

File: examples/end2end-example/java/src/test/java/runar/end2end/PriceBetTest.java
package runar.end2end;

import java.math.BigInteger;
import java.nio.file.Path;
import java.nio.file.Paths;

import org.junit.jupiter.api.Test;
import runar.lang.runtime.ContractSimulator;
import runar.lang.sdk.CompileCheck;
import runar.lang.types.Bigint;
import runar.lang.types.ByteString;
import runar.lang.types.PubKey;
import runar.lang.types.RabinPubKey;
import runar.lang.types.RabinSig;
import runar.lang.types.Sig;

import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertThrows;

/**
 * End-to-end PriceBet tests. Mirrors the TypeScript / Go / Python /
 * Ruby / Rust / Zig PriceBet test suites under
 * {@code examples/end2end-example/}.
 *
 * <p>Two layers of coverage:
 *
 * <ol>
 *   <li>Business logic via {@link ContractSimulator} (mocked Rabin/ECDSA
 *       crypto — every branch of {@code settle} and {@code cancel} is
 *       exercised).</li>
 *   <li>Compiler frontend via {@link CompileCheck} (the contract parses,
 *       validates, and typechecks through the real Rúnar pipeline).</li>
 * </ol>
 *
 * <p>The on-chain (regtest) flow lives in {@code integration/java} —
 * the dedicated PriceBet regtest case can be added there once a
 * Rabin-aware {@code IntegrationWallet} helper lands; this file
 * intentionally stays node-free so {@code gradle test} works in any
 * environment, matching the Go {@code PriceBet_test.go} pattern.
 */
class PriceBetTest {

    // Rabin test primes, both ≡ 3 (mod 4) so square roots are computable via
    // a^((p+1)/4) mod p. Same shared cross-tier test key as
    // runar.lang.runtime.RabinPaddingBoundTest and the examples/java
    // OraclePriceFeedTest. MockCrypto.verifyRabinSig performs REAL Rabin
    // verification (with the on-chain padding bound), so the oracle
    // signature must be honestly produced — garbage (sig, padding) pairs
    // no longer verify.
    private static final BigInteger P = new BigInteger("1361129467683753853853498429727072846227");
    private static final BigInteger Q = new BigInteger("1361129467683753853853498429727082846007");
    private static final BigInteger N = P.multiply(Q);

    private static final RabinPubKey ORACLE_PK = new RabinPubKey(N);

    private static final PubKey ALICE = PubKey.fromHex(
        "0202020202020202020202020202020202020202020202020202020202020202aa"
    );
    private static final PubKey BOB = PubKey.fromHex(
        "0303030303030303030303030303030303030303030303030303030303030303bb"
    );
    // 72-byte placeholder DER signatures — the simulator's checkSig is
    // mocked, so the bytes only have to be the right size.
    private static final Sig ALICE_SIG = Sig.fromHex("30440220" + "aa".repeat(32) + "0220" + "aa".repeat(32));
    private static final Sig BOB_SIG   = Sig.fromHex("30440220" + "bb".repeat(32) + "0220" + "bb".repeat(32));

    private static final Bigint STRIKE = Bigint.of(50_000);

    private static PriceBet makeBet() {
        return new PriceBet(ALICE, BOB, ORACLE_PK, STRIKE);
    }

    @Test
    void contractInstantiates() {
        PriceBet bet = makeBet();
        assertNotNull(bet);
    }

    @Test
    void settleAlicewinsWhenPriceExceedsStrike() {
        PriceBet bet = makeBet();
        ContractSimulator sim = ContractSimulator.stateless(bet);
        BigInteger[] sigAndPad = sign(priceMsg(60_000));
        sim.call("settle",
            Bigint.of(60_000), new RabinSig(sigAndPad[0]),
            new ByteString(unsignedToLE(sigAndPad[1])),
            ALICE_SIG, BOB_SIG
        );
    }

    @Test
    void settleBobWinsWhenPriceBelowStrike() {
        PriceBet bet = makeBet();
        ContractSimulator sim = ContractSimulator.stateless(bet);
        BigInteger[] sigAndPad = sign(priceMsg(30_000));
        sim.call("settle",
            Bigint.of(30_000), new RabinSig(sigAndPad[0]),
            new ByteString(unsignedToLE(sigAndPad[1])),
            ALICE_SIG, BOB_SIG
        );
    }

    @Test
    void settleBobWinsAtStrike() {
        // price == strike → bob wins (the contract's else branch).
        PriceBet bet = makeBet();
        ContractSimulator sim = ContractSimulator.stateless(bet);
        BigInteger[] sigAndPad = sign(priceMsg(50_000));
        sim.call("settle",
            STRIKE, new RabinSig(sigAndPad[0]),
            new ByteString(unsignedToLE(sigAndPad[1])),
            ALICE_SIG, BOB_SIG
        );
    }

    @Test
    void settleRejectsZeroPrice() {
        PriceBet bet = makeBet();
        ContractSimulator sim = ContractSimulator.stateless(bet);
        // Honestly signed zero price so the failure is attributable to the
        // price > 0 assert, not the signature layer.
        BigInteger[] sigAndPad = sign(priceMsg(0));
        assertThrows(AssertionError.class, () ->
            sim.call("settle",
                Bigint.of(0), new RabinSig(sigAndPad[0]),
                new ByteString(unsignedToLE(sigAndPad[1])),
                ALICE_SIG, BOB_SIG
            )
        );
    }

    @Test
    void settleRejectsForgedOracleSignature() {
        // Price exceeds the strike but the oracle signature is garbage —
        // the Rabin layer must reject it (regression for the stub-era
        // mock values this test previously relied on).
        PriceBet bet = makeBet();
        ContractSimulator sim = ContractSimulator.stateless(bet);
        assertThrows(AssertionError.class, () ->
            sim.call("settle",
                Bigint.of(60_000), new RabinSig(BigInteger.ONE),
                ByteString.fromHex("00"),
                ALICE_SIG, BOB_SIG
            )
        );
    }

    @Test
    void cancelSucceedsWithBothSignatures() {
        PriceBet bet = makeBet();
        ContractSimulator sim = ContractSimulator.stateless(bet);
        sim.call("cancel", ALICE_SIG, BOB_SIG);
    }

    @Test
    void compileCheckPasses() throws Exception {
        // Run the contract through the real Rúnar frontend (parse +
        // validate + typecheck) and confirm it produces a valid
        // artifact. Mirrors the Go TestPriceBet_Compile case.
        Path source = Paths.get(System.getProperty("user.dir"))
            .resolve("src/main/java/runar/end2end/PriceBet.runar.java");
        CompileCheck.run(source);
    }

    // --- honest signer (test-only; mirrors examples/java OraclePriceFeedTest) ---

    /** The message the contract verifies: num2bin(price, 8) — 8-byte little-endian. */
    private static byte[] priceMsg(long price) {
        byte[] msg = new byte[8];
        for (int i = 0; i < 8; i++) {
            msg[i] = (byte) (price >>> (8 * i));
        }
        return msg;
    }

    /** Returns {sig, padding} with padding < 65536 such that (sig^2 + padding) mod n == H(msg) mod n. */
    private static BigInteger[] sign(byte[] msg) {
        BigInteger hashBN = leToUnsigned(sha256(msg));
        BigInteger hashModN = hashBN.mod(N);

        for (long pad = 0; pad < 65536; pad++) {
            BigInteger padBig = BigInteger.valueOf(pad);
            BigInteger target = hashModN.subtract(padBig).mod(N);
            BigInteger root = sqrtModPQ(target);
            if (root != null) {
                BigInteger check = root.multiply(root).add(padBig).mod(N);
                if (check.equals(hashModN)) {
                    return new BigInteger[]{root, padBig};
                }
            }
        }
        throw new IllegalStateException("no valid padding < 65536 found for test message");
    }

    /** sqrt of a mod n=p*q via CRT, or null if not a quadratic residue. */
    private static BigInteger sqrtModPQ(BigInteger a) {
        BigInteger rp = sqrtMod3(a, P);
        if (rp == null) return null;
        BigInteger rq = sqrtMod3(a, Q);
        if (rq == null) return null;
        BigInteger qInvP = Q.modInverse(P);
        BigInteger pInvQ = P.modInverse(Q);
        BigInteger t1 = rp.multiply(Q).multiply(qInvP);
        BigInteger t2 = rq.multiply(P).multiply(pInvQ);
        return t1.add(t2).mod(N);
    }

    /** sqrt of a mod prime p where p ≡ 3 (mod 4); null if a is not a QR. */
    private static BigInteger sqrtMod3(BigInteger a, BigInteger p) {
        a = a.mod(p);
        if (a.signum() == 0) return BigInteger.ZERO;
        BigInteger r = a.modPow(p.add(BigInteger.ONE).shiftRight(2), p);
        return r.multiply(r).mod(p).equals(a) ? r : null;
    }

    private static byte[] sha256(byte[] data) {
        try {
            return java.security.MessageDigest.getInstance("SHA-256").digest(data);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    /** Unsigned little-endian bytes to BigInteger. */
    private static BigInteger leToUnsigned(byte[] le) {
        byte[] be = new byte[le.length];
        for (int i = 0; i < le.length; i++) be[le.length - 1 - i] = le[i];
        return new BigInteger(1, be);
    }

    /** BigInteger to minimal unsigned little-endian bytes. */
    private static byte[] unsignedToLE(BigInteger n) {
        if (n.signum() == 0) return new byte[]{0};
        byte[] be = n.toByteArray();           // big-endian, may have leading 0x00 sign byte
        int start = (be.length > 1 && be[0] == 0) ? 1 : 0;
        int len = be.length - start;
        byte[] le = new byte[len];
        for (int i = 0; i < len; i++) le[i] = be[be.length - 1 - i];
        return le;
    }
}
